src/OrbitGl/BatcherImpl.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_GL_BATCHER_IMPL_H_
+#define ORBIT_GL_BATCHER_IMPL_H_
+
+#include <Batcher.h>
+
+#include <array>
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "Containers/BlockChain.h"
+#include "PickingManager.h"
+
+struct LineBuffer {
+  void Reset() {
+    lines_.Reset();
+    colors_.Reset();
+    picking_colors_.Reset();
+  }
+
+  static const int NUM_LINES_PER_BLOCK = 64 * 1024;
+  orbit_containers::BlockChain<Line, NUM_LINES_PER_BLOCK> lines_;
+  orbit_containers::BlockChain<Color, 2 * NUM_LINES_PER_BLOCK> colors_;
+  orbit_containers::BlockChain<Color, 2 * NUM_LINES_PER_BLOCK> picking_colors_;
+};
+
+struct BoxBuffer {
+  void Reset() {
+    boxes_.Reset();
+    colors_.Reset();
+    picking_colors_.Reset();
+  }
+
+  static const int NUM_BOXES_PER_BLOCK = 64 * 1024;
+  orbit_containers::BlockChain<Box, NUM_BOXES_PER_BLOCK> boxes_;
+  orbit_containers::BlockChain<Color, 4 * NUM_BOXES_PER_BLOCK> colors_;
+  orbit_containers::BlockChain<Color, 4 * NUM_BOXES_PER_BLOCK> picking_colors_;
+};
+
+struct TriangleBuffer {
+  void Reset() {
+    triangles_.Reset();
+    colors_.Reset();
+    picking_colors_.Reset();
+  }
+
+  static const int NUM_TRIANGLES_PER_BLOCK = 64 * 1024;
+  orbit_containers::BlockChain<Triangle, NUM_TRIANGLES_PER_BLOCK> triangles_;
+  orbit_containers::BlockChain<Color, 3 * NUM_TRIANGLES_PER_BLOCK> colors_;
+  orbit_containers::BlockChain<Color, 3 * NUM_TRIANGLES_PER_BLOCK> picking_colors_;
+};
+
+struct PrimitiveBuffers {
+  void Reset() {
+    line_buffer.Reset();
+    box_buffer.Reset();
+    triangle_buffer.Reset();
+  }
+
+  LineBuffer line_buffer;
+  BoxBuffer box_buffer;
+  TriangleBuffer triangle_buffer;
+};
+
+/**
+Implements internal methods to collects primitives to be rendered at a later point in time.
+
+NOTE: The BatcherImpl assumes x/y coordinates are in pixels and will automatically round those
+down to the next integer in all Batcher::AddXXX methods. This fixes the issue of primitives
+"jumping" around when their coordinates are changed slightly.
+**/
+class BatcherImpl : public Batcher {
+ public:
+  BatcherImpl(BatcherId batcher_id, PickingManager* picking_manager = nullptr)
+      : Batcher(batcher_id, picking_manager) {}
+  [[nodiscard]] std::vector<float> GetLayers() const override;
+  void DrawLayer(float layer, bool picking) const override;
```

Comment:
After seeing the code, I could also imagine having multi level inheritance. Something like that:

- `Batcher` (inherits from nothing) declares a bunch of pure virtual functions for adding primitives
- `DrawingBatcher` (inherits from `Batcher`) declares `GetLayers` and `DrawLayer` as pure virtual functions
- `DrawingBatcherImpl` (inherits from `DrawingBatcher`) implements all of it. This class could be declared in a cpp-file and be a proper implementation detail.

For your tests you can mock or fake `Batcher`. The `CaptureWindow` would get a pointer to `DrawingBatcher` because it needs `DrawLayer`, but all the rest probably only needs a pointer to `Batcher`.

Reply:
Additional thoughts: You mention you want to rename this to `BatcherInterface`, but if I understand correctly, this will still include implementation of non-private methods? In that case it's not an interface. Did you consider composition over inheritance? For example:

* `BatcherInterface` declares only the virtual functions needed to add primitives, but as public API
* `OpenGlBatcher` implements this interface
* Create a new class `PrimitiveAssembler` (... better names welcome) that receives a pointer to a `BatcherInterface` and is responsible of assembling complex elements (circles, shaded boxes, ...) using the `BatcherInterface` methods

Reply:
Really thanks for the suggestions. A few thoughts. I feel:

- The goal of this PR is isolate OpenGl related methods with the rest in Batcher (Thanks Florian for the better name, I should use OpenGlBatcher!).
- Henning is trying to additionally isolate drawing primitives with adding them.
- Florian is trying to additionally isolate proper primitives with additional methods to add them. 

I found both proposals correct, I was a bit lazy to think about all this refactors in this PR, but I should have a clearly defined long-time proposal before merging this, thanks for making me realize that.

If we are worried about the long-term picture, I will include merging TextRenderer and Batcher. For example, in Florian's proposal, PrimitiveAssembler could have both pointer of BatcherInterface and TextRendererInterface.

I will update the doc before going forward with this PR.

Reply:
Fair enough to not do this all in a single PR, this can absolutely be approached iteratively and I think this is the first step in the right direction.